analyzers/src/SonarAnalyzer.Common/Rules/StringConcatenationInLoopBase.cs

```diff
@@ -20,130 +20,80 @@
 
 namespace SonarAnalyzer.Rules
 {
-    public abstract class StringConcatenationInLoopBase : SonarDiagnosticAnalyzer
+    public abstract class StringConcatenationInLoopBase<TSyntaxKind, TAssignmentExpression, TBinaryExpression> : SonarDiagnosticAnalyzer<TSyntaxKind>
+        where TSyntaxKind : struct
+        where TAssignmentExpression : SyntaxNode
+        where TBinaryExpression : SyntaxNode
     {
         protected const string DiagnosticId = "S1643";
-        protected const string MessageFormat = "Use a StringBuilder instead.";
+        protected override string MessageFormat => "Use a StringBuilder instead.";
 
-        protected abstract GeneratedCodeRecognizer GeneratedCodeRecognizer { get; }
-    }
+        protected abstract TSyntaxKind[] CompoundAssignmentKinds { get; }
+        protected abstract ISet<TSyntaxKind> ExpressionConcatenationKinds { get; }
+        protected abstract ISet<TSyntaxKind> LoopKinds { get; }
+        protected virtual bool IsAddExpression(TBinaryExpression expression) => true;
```

Comment:
Should be abstract. VB also have a lot of different BinaryExpressions:
https://learn.microsoft.com/en-us/dotnet/api/microsoft.codeanalysis.visualbasic.syntax.binaryexpressionsyntax?view=roslyn-dotnet-4.6.0